File: ObjC/SIALogger/SIALogger/Sources/LoggerOutput/SIALogDocumentsFileOutput.m
//
//  SIALogDocumentsFileOutput.m
//  SIALogger
//
//  Created by Ivlev  Alexander (Stef) on 4/15/15.
//  Copyright (c) 2016 Ivlev Alexander. All rights reserved.
//

#import "SIALogDocumentsFileOutput.h"

@interface SIALogDocumentsFileOutput ()

@property (nonatomic, strong) NSFileHandle* output;

@end

@implementation SIALogDocumentsFileOutput

+ (NSString*)documentsDirectory {
  NSArray<NSString*>* paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
  NSAssert(paths.count > 0, @"Can't search path for documetn directory.");
  
  return [paths objectAtIndex:0];
}

+ (NSString*)pathWithFileName:(NSString*)fileName {
  return [NSString stringWithFormat:@"%@/%@.log", [self documentsDirectory], fileName];
}

+ (NSString*)pathWithDateAndFileName:(NSString*)fileName {
  NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
  [dateFormatter setDateFormat:@"yyyyMMdd_HHmmss"];
  [dateFormatter setTimeZone:[NSTimeZone timeZoneWithName:@"UTC"]];

  NSString* dateString = [dateFormatter stringFromDate:[NSDate date]];

  return [NSString stringWithFormat:@"%@/%@_%@.log", [self documentsDirectory], fileName, dateString];
}

- (instancetype)initWithFileName:(NSString*)fileName joinDate:(BOOL)join {
  self = [super init];
  if (self) {
    NSString* filePath = join ?
      [SIALogDocumentsFileOutput pathWithDateAndFileName:fileName] :
      [SIALogDocumentsFileOutput pathWithFileName:fileName];

    [[NSFileManager defaultManager] createFileAtPath:filePath contents:nil attributes:nil];
    self.output = [NSFileHandle fileHandleForWritingAtPath:filePath];
    
    NSAssert(nil != self.output, @"Can't create file handler for path:%@", filePath);
  }

  return self;
}

- (void)logWithTime:(NSString*)time Level:(SIALogLevel*)level File:(NSString*)file Line:(NSNumber*)line Msg:(NSString*)msg {
  assert(nil != time && nil != level && nil != file && nil != line && nil != msg);
  
  NSString* log = [NSString stringWithFormat:@"%@ [%@] {%@:%@}: %@\r\n", time, level.name.uppercaseString, file, line, msg];
  [self.output writeData:[log dataUsingEncoding:NSUTF8StringEncoding]];
}

@end

File: Swift/SIALogger/SIALogger/LoggerOutput/SIALogDocumentsFileOutput.swift
//
//  SIALogDocumentsFileOutput.swift
//  SIALogger
//
//  Created by Alexander Ivlev on 31/05/16.
//  Copyright © 2016 Alexander Ivlev. All rights reserved.
//

import Foundation

public class SIALogDocumentsFileOutput : SIALogOutputProtocol {
  public required init?(fileName: String, joinDate: Bool = true) {
    guard let filePath = SIALogDocumentsFileOutput.createFilePath(fileName, joinDate: joinDate) else {
      return nil
    }
    
    NSFileManager.defaultManager().createFileAtPath(filePath, contents: nil, attributes: nil)
    
    guard let outputHandle = NSFileHandle(forWritingAtPath: filePath) else {
      return nil
    }
    self.outputHandle = outputHandle
  }
  
  public func log(message: String) {
    if let data = (currentTime()+message).dataUsingEncoding(NSUTF8StringEncoding) {
      self.outputHandle.writeData(data)
    }
  }
  
  
  private let outputHandle : NSFileHandle
  private let startDate = NSDate()
  
  private func currentTime() -> String {
    let dateFormatter = NSDateFormatter()
    
    dateFormatter.dateFormat = "HH:mm:ss:SSS"
    dateFormatter.timeZone = NSTimeZone(name: "UTC")
    
    return dateFormatter.stringFromDate(NSDate(timeInterval: 0, sinceDate: startDate))
  }
  
  private static func createFilePath(fileName: String, joinDate: Bool) -> String? {
    guard let documentDirectory = SIALogDocumentsFileOutput.documentsDirectory() else {
      return nil
    }
    
    let dateFormatter = NSDateFormatter()
    dateFormatter.dateFormat = "yyyyMMdd_HHmmss";
    dateFormatter.timeZone = NSTimeZone(name: "UTC")
    
    return documentDirectory+"/"+fileName+"_"+dateFormatter.stringFromDate(NSDate())+".log"
  }
  
  private static func documentsDirectory() -> String? {
    let directory = NSSearchPathDirectory.DocumentDirectory;
    let domainMask = NSSearchPathDomainMask.AllDomainsMask;
    
    guard let dir = NSSearchPathForDirectoriesInDomains(directory, domainMask, true).first else {
      return nil
    }
    return dir
  }
}